src/cargo/util/toml/mod.rs

```diff
@@ -2057,6 +2092,16 @@ impl TomlTarget {
         }
     }
 
+    fn validate_proc_macro(&self, warnings: &mut Vec<String>) {
+        if self.proc_macro_raw.is_some() && self.proc_macro_raw2.is_some() {
+            warnings.push(format!(
+                "found both `proc-macro` and `proc_macro` are set \
+                 in the `{}` library target",
+                self.name()
+            ));
+        }
+    }
+
```

Comment:
I feel like `TomlManifest` is more like a readonly struct and might not want these validation method on it. How about moving `validate_proc_macro` and `validate_crate_types` into `util::toml::target`, just like how `validate_target_name` works?

Reply:
This verification method was learned from TomlProfiles.